docs/diplomat-and-macros.md

```diff
@@ -0,0 +1,272 @@
+and while it is built from a `syn` struct, the final representation is independent of `syn`
+and its ast representation.
+
+## UniFFI's experience with the macro approach.
+
+Ryan tried this same macro approach for UniFFI in [#416](https://github.com/mozilla/uniffi-rs/pull/416) -
+but we struck **what's bad about this** - at least for UniFFI's use-cases: the context in which the
+macro runs doesn't know about types defined outside of that macro, which are what we need to
+expose.
+
+### Example of these limitations
+
+Let's look at diplomat's simple example:
+
+```rust
+#[diplomat::bridge]
+mod ffi {
+    pub struct MyFFIType {
+        pub a: i32,
+        pub b: bool,
+    }
+
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+This works fine, but starts to come unstuck if you want the types defined somewhere else. In this trivial example, something like:
+
+```Rust
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+
+#[diplomat::bridge]
+mod ffi {
+    impl MyFFIType {
+        pub fn create() -> MyFFIType { ... }
+        ...
+    }
+}
+```
+
+fails - diplomat can't handle this scenario - in the same way and for the same reasons that Ryan's
+[#416](https://github.com/mozilla/uniffi-rs/pull/416) can't - the contents of the struct aren't known.
+
+From the Rust side of the world, this is probably solvable by sprinkling more macros around - eg, something like:
+
+```Rust
+#[uniffi::magic]
+pub struct MyFFIType {
+    pub a: i32,
+    pub b: bool,
+}
+```
+
+Might be enough for the generation of the Rust scaffolding - in UniFFI's case, all we really need
+is an implementation of `uniffi::RustBufferViaFfi` which is easy to derive, and UniFFI can
+generate code which assumes that exists, much like it does now.
+However, the problems are in the foreign bindings, because, eg, those foreign bindings do not know
+the names and types of the struct elements without re-parsing every bit of Rust code with those
+annotations. As discussed below, re-parsing this code might be an option if we help Uniffi to
+find it, but asking UniFFI to parse this and all dependent crates to auto-discover them
+probably is not going to be viable.
+
+### Why is this considered a limition for UniFFI but not diplomat?
+
+As mentioned above, diplomat considers the limitation described above as an intentional design
+feature. By limiting where FFI types can be described, there's no risk of changes made "far away"
+from the FFI to change the FFI. This was born of experience in tools like `cbindgen`.
+
+For Uniffi, all use-cases needed by Mozilla don't share this design goal, primarily because the
+FFI is the primary consumer of the crate. The Rust API exists purely to service the FFI. It's not
+really possible to accidentally change the API, because every API change made will be in service
+of exposing that change over the FFI. The test suites written in the foreign languages are
+considered canonical.
```

Comment:
Ah, I think this helps answer my question about about API stability in UniFFI vs Diplomat. I still feel like the `.udl` serves a similar purpose in the UniFFI world, but because of the way the `.udl` kind of maps directly onto the underlying Rust code, I can see an argument that changes in a UniFFI component are *more likely* to change the FFI surface than in Diplomat.

They won't change to FFI surface by stealth, because you'll have to update the `.udl` file...but changing the `.udl` file is likely to be the simplest way to accommodate a change in the underlying Rust code. In Diplomat you might instead write some adapter code in the FFI module in order to avoid changing the FFI surface, because there's an obvious place to do it.

Reply:
yeah, I think we agree that UniFFI's use of the .udl file and Diplomat's decision to restrict where types are exposed do serve the same purpose in that regard. The broader point I'm trying to make though is that application-services would prefer to *not* have those guards in place - ie, it would probably prefer a world where the UDL file didn't exist and nor did any limitation about where these types could be defined.